message_ix_models/tools/iea/weo.py

```diff
@@ -0,0 +1,672 @@
+        "solarpv_buildings": ["Renewables", 15],
+        "wind_onshore": ["Renewables", 25],
+        "wind_offshore": ["Renewables", 35],
+        "hydropower_large": ["Renewables", 45],
+        "hydropower_small": ["Renewables", 55],
+        "bioenergy_large": ["Renewables", 65],
+        "bioenergy_cofiring": ["Renewables", 75],
+        "bioenergy_medium_chp": ["Renewables", 85],
+        "bioenergy_ccus": ["Renewables", 95],
+        "csp": ["Renewables", 105],
+        "geothermal": ["Renewables", 115],
+        "marine": ["Renewables", 125],
+    }
+
+    # Specify cost types to read in and the required columns
+    cost_cols = {"capital_costs": "A,B:D", "annual_om_costs": "A,F:H"}
+
+    # Loop through each technology and cost type
+    # Read in data and convert to long format
+    dfs_cost = []
+    for tech_key in tech_rows:
+        for cost_key in cost_cols:
+            df = pd.read_excel(
+                file_path,
+                sheet_name=tech_rows[tech_key][0],
+                header=None,
+                skiprows=tech_rows[tech_key][1],
+                nrows=8,
+                usecols=cost_cols[cost_key],
+            )
+
+            df.columns = ["region", "2021", "2030", "2050"]
+            df_long = pd.melt(
+                df, id_vars=["region"], var_name="year", value_name="value"
+            )
+
+            df_long["scenario"] = "stated_policies"
+            df_long["technology"] = tech_key
+            df_long["cost_type"] = cost_key
+            df_long["units"] = "usd_per_kw"
+
+            # Reorganize columns
+            df_long = df_long[
+                [
+                    "scenario",
+                    "technology",
+                    "region",
+                    "year",
+                    "cost_type",
+                    "units",
+                    "value",
+                ]
+            ]
+
+            dfs_cost.append(df_long)
+
+    all_cost_df = pd.concat(dfs_cost)
+
+    return all_cost_df
+
+
+"""
+Match each R11 region with a WEO region
+"""
+dict_weo_r11 = {
+    "NAM": "United States",
+    "LAM": "Brazil",
+    "WEU": "European Union",
+    "EEU": "Russia",
+    "FSU": "Russia",
+    "AFR": "Africa",
+    "MEA": "Middle East",
+    "SAS": "India",
+    "CPA": "China",
+    "PAS": "India",
+    "PAO": "Japan",
+}
+
+
+def calculate_region_cost_ratios(weo_df, dict_reg):
+    """
+    Returns DataFrame of cost ratios (investment cost and O&M cost) for each R11 region,
+    for each technology
+
+    Only returns values for the earliest year in the dataset
+    (which, as of writing, is 2021)
+    """
+
+    # Replace "n.a." strings with NaNs
+    weo_df["value"] = weo_df["value"].replace("n.a.", np.nan)
+
+    # Filter for only United States data (this is the NAM region)
+    df_us = weo_df.loc[weo_df.region == "United States"].copy()
+
+    # Rename the `value` column in the US dataframe to `us_value`
+    df_us.rename(columns={"value": "us_value"}, inplace=True)
```

Comment:
Any time you notice yourself using `inplace=True`, especially successively, can be replaced with chaining.